rfcs/004 - Continuous Two-Way Synchronization RFC.md

```diff
@@ -0,0 +1,160 @@
+4. The database, in an eventually-consistent way, always reflects the state of your cloud, so querying is straightforward.
+5. Two-way mode auto-importing resources created by "other" things should make auto-generating limited modules possible (create an entity from a `List-` AWS API call, then turn every other AWS API into an RPC call from the database, allowing imperative Create/Update/Delete and a read-only table to SELECT results from that is re-synced with the cloud every so often), which will let us more rapidly hit 100% cloud coverage (though with a second-tier of quality), which can help us with larger startups and enterprises that need full coverage more than they need cleaner, mutable tables with explicit dependency foreign keys.
+
+That last piece is particularly interesting because it's a way to help us accelerate our development velocity and potentially turn our current "low-level modules" into "mid-level modules" that could be pure SQL modules depending on these RPC functions, and the "high-level modules" could decide to depend on either of them, depending on what is easier.
+
+### Alternatives Considered
+
+#### Just don't implement any `code-` modules
+
+Basically, we can ~~stick our heads in the sand~~ decide to enforce management of all cloud resources in a way that these changes are represented in our database, but this makes our ability to co-exist with other IaC tools impossible, as well as integrating with any sort of CI system (without basically re-implementing our own CI system we then demand users use). And therefore this approach has been rejected.
+
+#### Implement `iasql_commit` as replacement for `iasql_apply`/`iasql_sync` only
+
+Stick to manual mode and just replace `apply` and `sync` with `iasql_commit`. This allows us to implement the `code-` modules and requires less work, but keeps the harder to explain syntax in the forefront, rather than being a power user tool. It also precludes our ability to support Enforce/Track behavior described above, which narrows our set of advantages.
+
+This may still be a point along-the-way in the conversion to the Full Two-Way Mode, though, but has been rejected as the stopping point.
+
+#### Implement Two-Way Mode with the transaction-like syntax only
+
+Basically just the first half of the proposal, stopping before the Enforce/Track section. This is probably also a point along-the-way, but has been rejected as the stopping point because the remaining features are simpler to implement and provide significant functionality.
+
+## Expected Semver Impact
+
+If we were post v1.0.0, this would be a major update.
+
+## Affected Components
+
+- The `iasql_functions` module
+- The core RPC scheduler
+- The entire test suite
+- Most of the documentation
+
+## Expected Timeline
+
+This will take a few weeks to get across the line. The major steps (with estimated times) are:
```

Comment:
I'm missing here (and in the RFC) the audit logs. I think they are the foundational part for this RFC to work properly and maybe need a paragraph on how they are expected to work or look and its own timeline in this section. It will not only be the table creation but also add the triggers for all tables.

Reply:
So we already have audit logs implemented. We simply need to use them. I think it should simply be a collection of select queries that we use, potentially joining in the user group/role onto the table, but I not sure that is necessary.

I can expand this with example queries that I am thinking of here.